clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp

```diff
@@ -365,10 +365,15 @@ mlir::Attribute ConstantEmitter::tryEmitPrivateForVarInit(const VarDecl &d) {
   if (!d.hasLocalStorage()) {
     QualType ty = cgm.getASTContext().getBaseElementType(d.getType());
     if (ty->isRecordType())
-      if (d.getInit() && isa<CXXConstructExpr>(d.getInit())) {
-        cgm.errorNYI(d.getInit()->getBeginLoc(),
-                     "tryEmitPrivateForVarInit CXXConstructExpr");
-        return {};
+      if (const CXXConstructExpr *e =
+              dyn_cast_or_null<CXXConstructExpr>(d.getInit())) {
+        const CXXConstructorDecl *cd = e->getConstructor();
```

Comment:
Does classic code-gen assume that `cd` is non-null here?  I'm not positive it always is.

Also, do we/should we model elidable here?

Reply:
Ok, glad to know.  Sometimes the `decl` elements of an expr can be null in subtle situations.

As far as elidable.  There is a `CXXConstructExpr::isElidable` that I'm surprised we don't have to express here, since the idea is that we could skip the construction.

Reply:
I see. If I add `-std=c++14` to my compile line, that gets the elidable constructor in the AST, and then `-fno-elide-constructors` get us to emit CIR for it but not here. That's handled in `emitCXXConstructExpr` (which has not been upstreamed yet). I guess the `cd->isTrivial() && cd->isDefaultConstructor()` check prevents us from doing anything with it here.

Reply:
We have been careful not to elide things too early since it might hurt static analysis - this is a big pain point and request from the static analysis community (as clang really early get rid of these). The idea is to represent them in a way that during LoweringPrepare we can either elide them out or apply specific lowering heuristics (e.g. when deciding the best way to initialize large set of constant datas).

Reply:
@bcardosolopes Does that mean that we should be emitting an elidable constructor in CIR even if the `-fno-elide-constructors` option isn't used? The incubator is eliding them the same way classic codegen does.

For the case without `-std=c++14` I'm not sure we have much choice since it's not even in the AST.